merge.py

```diff
@@ -1,98 +1,303 @@
+
+            tmp_dir = TMP_DIR_FMT.format(dir='{}-{}'.format(
+                self.config['repository'], pr_num))
+            try:
+                _set_up(tmp_dir)
+            except AssertionError:
+                pr.post_error('Setup of temp directory failed, try again.')
+                continue
+            if self.merge_git_pr(pr, tmp_dir, m_l):
+                m_l.info('Merge concluded satisfactorily. Moving on.')
+                self.main_logger.info(
+                    'PR#{num} processing done.'.format(num=pr_num))
+            else:
+                m_l.info('Merge did not conclude satisfactorily (reporting to'
+                         ' github failed). Adding PR back to queue to be tried'
+                         ' again.')
+                self.main_logger.info(
+                    'PR#{num} processing failed.'.format(num=pr_num))
+                self.work_queue.put(pr)
+            _clean_up(tmp_dir)
+
+    def get_logger(self, pr_num):
+        """get_logger returns a logger for a particular PR.
+
+        Args:
+            pr_num: The pull request number.
+        Returns:
+            logger configured to output to an appropriate file.
+        """
+        output_file = '{name}_pr_{pr_num}_merge_log.txt'.format(
+            name=self.config['name'], pr_num=pr_num)
+        m_l = logging.getLogger('{name}_{num}_merge_logger'.format(
+            name=self.config['name'], num=pr_num))
+        if not m_l.handlers:
+            log_fmt = '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
+            f = logging.Formatter(log_fmt)
+            h = logging.FileHandler(os.path.join('log', output_file))
+            h.setFormatter(f)
+            m_l.addHandler(h)
+            m_l.setLevel(logging.INFO)
+        return m_l
+
+    def merge_git_pr(self, pr, tmp_dir, m_l):
+        """merge_git_pr merges git pull requests.
+
+        Args:
+            pr: The GithubPR to merge.
+            tmp_dir: The directory in which to do work.
+            m_l: merge logger
+        Returns:
+            True if merge was concluded satisfactorily (merged successfully,
+            or failed due to supposed fault of the PR itself).
+            False if failure was due to an environmental issue and should be
+            retried.
+        """
+        branch = self.config['merge_branch']
+        org = self.config['github_org']
+        remote_name = 'apache'
+        repo = self.config['repository']
+        pr_vars = {
+            'apache_url': APACHE_GIT.format(repo=repo),
+            'branch': branch,
+            'branch_path': '{}/{}'.format(remote_name, branch),
+            'msg': 'This closes #{}'.format(pr.get_num()),
+            'remote_name': remote_name,
+            'repo': repo,
+            'repo_url': GITHUB_REPO_URL.format(org=org, repo=repo),
+            'pr_name': 'finish-pr-{}'.format(pr.get_num()),
+            'pr_num': pr.get_num(),
+        }
+        try:
+            cmds = [
```

Comment:
I'd suggest putting these commands in a constant somewhere -- they should only depend on the contents of the pr_vars. So maybe something like:

```
# Commands to run for merging and validating a PR. The following variables will be available
# in the environment:
#   apache_url: The URL of the Apache repository
#   branch: The branch being merged to
#   ...  
MERGE_COMMANDS = [
  ...
]
```

Then this method is just:
```
try:
  self.run_commands(CMDS, pr_vars, ...);
except ...
   ...
```